File: app/src/main/java/com/drdisagree/iconify/ui/utils/TaskExecutor.java
package com.drdisagree.iconify.ui.utils;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public abstract class TaskExecutor<Params, Progress, Result> {

    private final Handler mHandler;
    private final Executor mExecutor;
    private final AtomicBoolean mCancelled;
    private final CountDownLatch preExecuteLatch;
    private volatile Status mStatus;

    public TaskExecutor() {
        mExecutor = Executors.newSingleThreadExecutor();
        mHandler = new Handler(Looper.getMainLooper());
        mStatus = Status.PENDING;
        mCancelled = new AtomicBoolean(false);
        preExecuteLatch = new CountDownLatch(1);
    }

    public final Status getStatus() {
        return mStatus;
    }

    @SafeVarargs
    public final void execute(final Params... params) {
        if (mStatus != Status.PENDING) {
            switch (mStatus) {
                case RUNNING:
                    throw new IllegalStateException("Cannot execute task: the task is already running.");
                case FINISHED:
                    throw new IllegalStateException("Cannot execute task: the task has already been executed (a task can be executed only once)");
            }
        }

        mExecutor.execute(() -> {
            mStatus = Status.RUNNING;

            try {
                mHandler.post(() -> {
                    if (!isCancelled()) {
                        onPreExecute();
                    }
                    preExecuteLatch.countDown();
                });

                preExecuteLatch.await();

                if (!isCancelled()) {
                    final Result result = runInBackground(params);

                    mHandler.post(() -> {
                        if (!isCancelled()) {
                            onPostExecute(result);
                        }
                    });
                }
            } catch (Throwable throwable) {
                mCancelled.set(true);

                try {
                    throw throwable;
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            } finally {
                if (isCancelled()) {
                    onCancelled();
                }

                mStatus = Status.FINISHED;
            }
        });
    }

    protected abstract void onPreExecute();

    @SafeVarargs
    private final Result runInBackground(Params... params) {
        if (!isCancelled()) {
            return doInBackground(params);
        }
        return null;
    }

    @SuppressWarnings({"unchecked", "unsafe"})
    protected abstract Result doInBackground(Params... params);

    @SafeVarargs
    protected final void publishProgress(Progress... values) {
        if (!isCancelled()) {
            mHandler.post(() -> onProgressUpdate(values));
        }
    }

    @SuppressWarnings({"unchecked", "unsafe"})
    protected void onProgressUpdate(Progress... values) {
    }

    protected abstract void onPostExecute(Result result);

    protected void onCancelled() {
    }

    public final boolean isCancelled() {
        return mCancelled.get();
    }

    public final void cancel(boolean mayInterruptIfRunning) {
        mCancelled.set(true);
        if (mayInterruptIfRunning) {
            mHandler.removeCallbacksAndMessages(null);
        }
    }

    public enum Status {
        PENDING,
        RUNNING,
        FINISHED,
    }
}

File: app/src/main/java/com/drdisagree/iconify/utils/extension/TaskExecutor.kt
package com.drdisagree.iconify.utils.extension

import android.os.Handler
import android.os.Looper
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.concurrent.Volatile

abstract class TaskExecutor<Params, Progress, Result> {

    private val mHandler: Handler = Handler(Looper.getMainLooper())
    private val mExecutor: Executor = Executors.newSingleThreadExecutor()
    private val mCancelled: AtomicBoolean
    private val preExecuteLatch: CountDownLatch
    private val isCancelled: Boolean
        get() = mCancelled.get()

    @Volatile
    var status: Status
        private set

    init {
        status = Status.PENDING
        mCancelled = AtomicBoolean(false)
        preExecuteLatch = CountDownLatch(1)
    }

    @SafeVarargs
    fun execute(vararg params: Params) {
        if (status != Status.PENDING) {
            when (status) {
                Status.RUNNING -> throw IllegalStateException("Cannot execute task: the task is already running.")
                Status.FINISHED -> throw IllegalStateException("Cannot execute task: the task has already been executed (a task can be executed only once)")
                Status.PENDING -> {}
            }
        }

        mExecutor.execute {
            status = Status.RUNNING

            try {
                mHandler.post {
                    if (!isCancelled) {
                        onPreExecute()
                    }

                    preExecuteLatch.countDown()
                }

                preExecuteLatch.await()

                if (!isCancelled) {
                    val result = runInBackground(*params)

                    mHandler.post {
                        if (!isCancelled) {
                            onPostExecute(result)
                        }
                    }
                }
            } catch (throwable: Throwable) {
                mCancelled.set(true)

                try {
                    throw throwable
                } catch (e: InterruptedException) {
                    throw RuntimeException(e)
                }
            } finally {
                if (isCancelled) {
                    onCancelled()
                }

                status = Status.FINISHED
            }
        }
    }

    protected abstract fun onPreExecute()

    @SafeVarargs
    private fun runInBackground(vararg params: Params): Result? {
        return if (!isCancelled) {
            doInBackground(*params)
        } else null
    }

    protected abstract fun doInBackground(vararg params: Params): Result

    @SafeVarargs
    protected fun publishProgress(vararg values: Progress) {
        if (!isCancelled) {
            mHandler.post { onProgressUpdate(*values) }
        }
    }

    protected open fun onProgressUpdate(vararg values: Progress) {}
    protected abstract fun onPostExecute(result: Result?)
    protected open fun onCancelled() {}

    fun cancel(mayInterruptIfRunning: Boolean) {
        mCancelled.set(true)
        if (mayInterruptIfRunning) {
            mHandler.removeCallbacksAndMessages(null)
        }
    }

    enum class Status {
        PENDING,
        RUNNING,
        FINISHED
    }
}
